File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.js
/**
 * @param {character[][]} maze
 * @param {number[]} entrance
 * @return {number}
 */
var nearestExit = function(maze, entrance) {
  const [m, n] = [maze.length, maze[0].length];
  const dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]];
  const queue = [[...entrance, 0]];
  
  while (queue.length) {
    const [r, c, cost] = queue.shift();
    if ((r !== entrance[0] || c !== entrance[1]) && 
       (!r || r === m - 1 || !c || c === n - 1)) return cost;
    
    for (const [dx, dy] of dirs) {
      const [x, y] = [dx + r, dy + c];
      if (x < 0 || x >= m || y < 0 || y >= n || maze[x][y] !== '.') continue;
      
      maze[x][y] = '+';
      queue.push([x, y, cost + 1]);
    }
  }
  
  return -1;
};

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        m, n = len(maze), len(maze[0])
        dq = deque([entrance + [0]])
        while dq:
            r, c, cost = dq.popleft()
            if (r != entrance[0] or c != entrance[1]) and (not r or r == m - 1 or not c or c == n - 1):
                return cost
            for dx, dy in dirs:
                x, y = dx + r, dy + c
                if x < 0 or x >= m or y < 0 or y >= n or maze[x][y] != '.':
                    continue
                maze[x][y] = '+'
                dq.append([x, y, cost + 1])
        return -1
        
